File: codes/algo/backtrace/P1021连续邮资.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <limits>
#include <functional>

using namespace std;

int max_covers(const vector<int>& stamps, int N) {
    int stamp_type = stamps.size();
    int max_value = (stamp_type > 0 ? stamps.back() : 0) * N;
    if (max_value == 0) return 0;

    vector<vector<int>> min_use(stamp_type, vector<int>(max_value + 1, numeric_limits<int>::max()));

    for (int j = 0; j <= max_value; ++j) {
        if (j > N) break;
        min_use[0][j] = j;
    }

    for (int i = 0; i < stamp_type; ++i) {
        min_use[i][0] = 0;
    }

    for (int i = 1; i < stamp_type; ++i) {
        for (int j = 1; j <= max_value; ++j) {
            min_use[i][j] = min_use[i-1][j];
            for (int t = 1; t <= N; ++t) {
                if (j < t * stamps[i]) break;
                if (j - t * stamps[i] >= 0 && min_use[i - 1][j - t * stamps[i]] != numeric_limits<int>::max()) {
                    min_use[i][j] = min(min_use[i][j], t + min_use[i - 1][j - t * stamps[i]]);
                }
            }
        }
    }

    int max_reach = 0;
    for (int j = 1; j <= max_value; ++j) {
        if (min_use[stamp_type - 1][j] <= N) {
            max_reach = j;
        } else {
            break;
        }
    }
    return max_reach;
}

int main() {
    int N, K;
    cin >> N >> K;

    vector<int> path = {1};
    int ans = 1;
    vector<int> stamps;

    function<void(int)> backtrace = [&](int i) {
        if (i == K) {
            int r = max_covers(path, N);
            if (r > ans) {
                ans = r;
                stamps = path;
            }
            return;
        }

        for (int v = path.back() + 1; v <= ans + 2; ++v) {
            path.push_back(v);
            backtrace(i + 1);
            path.pop_back();
        }
    };

    backtrace(1);

    for (size_t i = 0; i < stamps.size(); ++i) {
        cout << stamps[i] << (i == stamps.size() - 1 ? "" : " ");
    }
    cout << endl;
    cout << "MAX=" << ans << endl;

    return 0;
}

File: codes/algo/backtrace/P1021连续邮资.py
from math import inf


def max_covers(stamps: list, N: int):
    """stamps这个邮票组合,能够达到的连续邮资的最大值r"""
    stamp_type = len(stamps)
    max_value = stamps[-1] * N  # 理论能达到的最大面值
    min_use = [
        [inf] * (max_value + 1) for _ in range(stamp_type)
    ]  # min_use[i][j] 表示使用i前面的 邮票到达j面额，所需要的最小张数

    for j in range(max_value):
        if j * 1 > N:
            break
        min_use[0][j] = j

    for i in range(stamp_type):
        min_use[i][0] = 0

    for i in range(stamp_type):
        for j in range(1, max_value):
            cur_min = min_use[i][j]
            for t in range(N):
                if j < t * stamps[i]:
                    break
                cur_min = min(cur_min, t + min_use[i - 1][j - t * stamps[i]])
            min_use[i][j] = cur_min

    # for x in min_use:
    #     print(x)

    # r = [0] * stamp_type

    # for i in range(stamp_type):
    #     for j in range(max_value):
    #         if min_use[i][j + 1] > N and min_use[i][j] <= N:
    #             r[i] = j
    # return r
    max_reach = 0
    for j in range(1, max_value + 1):
        if min_use[stamp_type - 1][j] <= N:
            max_reach = j
        else:
            break
    return max_reach


# test
# ans = max_covers([1, 3], 3)
# print(ans)

N, K = map(int, input().split())

path = [1]
ans = 1
stamps = []


def backtrace(i):
    global ans, stamps
    r = max_covers(path, N)
    if i == K:
        if r >= ans:
            ans = r
            stamps = path.copy()
        return

    for v in range(path[-1] + 1, r + 2):
        path.append(v)
        backtrace(i+1)
        path.pop()
    

backtrace(1)
print(" ".join([str(x) for x in stamps]))
print(f"MAX={ans}")